feat(backend): configurable log level for driver / launcher images

CarterFendley opened the pull request:
### IMPORTANT
Design changes have been made since the original description, see [this comment](https://github.com/kubeflow/pipelines/pull/11278#issuecomment-2670098661) for updated usage


---
---

### Description of your changes:
This PR adds the ability to change log level in the driver / launcher containers. This is implemented in a similar pattern as the overrides for driver / launcher images. Specifically, you can add the following environment variables to the `ml-pipeline` deployments:

```yaml
    spec:
      containers:
      - env:
        - name: DRIVER_LOG_LEVEL
          value: "3"
        - name: LAUNCHER_LOG_LEVEL
          value: "3"
```

**Note:** A numerical value such as the literal `3` not `"3"` here will be invalid deployment spec and validation on the spec will fail causing `kubectl edit` to reject it with the message: `error: deployments.apps "ml-pipeline" is invalid`.

#### Other minor alterations

1. In [this commit](https://github.com/kubeflow/pipelines/commit/76cd57179f6bc50ead5c2cce538295ebc54b8b14) two locations were updated to use the `workflowCompiler.driverImage` and `workflowCompiler.launcherImage` attributes which are populated [here](https://github.com/kubeflow/pipelines/blob/76cd57179f6bc50ead5c2cce538295ebc54b8b14/backend/src/v2/compiler/argocompiler/argo.go#L126-L127). **This is a very minor change** but seemed better to invoke only once and match other such usages (in [importer.go](https://github.com/kubeflow/pipelines/blob/76cd57179f6bc50ead5c2cce538295ebc54b8b14/backend/src/v2/compiler/argocompiler/importer.go#L95) and [dag.go](https://github.com/kubeflow/pipelines/blob/76cd57179f6bc50ead5c2cce538295ebc54b8b14/backend/src/v2/compiler/argocompiler/dag.go#L429)). If there are reasons this **should** be re-invoked, please let me know.
2. The [--copy flag](https://github.com/kubeflow/pipelines/blob/8590ec1147012f4a7c95f594d71f290f566bc0e8/backend/src/v2/compiler/argocompiler/container.go#L307) were moved [into the arguments](https://github.com/kubeflow/pipelines/blob/285cb049f14706ea4e1775a0a99ff2ded2ad22a4/backend/src/v2/compiler/argocompiler/container.go#L330) block, to match other implementations. Again, lmk if this is not wanted.

### Feedback wanted

The environment variable for this is similar to the `V2_LAUNCHER_IMAGE` and `V2_DRIVER_IMAGE` but without the `V2_` prefix. If anyone has preferences here, I do not, so happy to take any path. 

### Checklist:
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

hbelmiro commented:
/ok-to-test

droctothorpe reviewed (approved):
LGTM. Thank you for tackling this, @CarterFendley!

rimolive commented:
Maybe worth adding just one unit test to verify if setting both env vars will generate a IR yaml with the new flags.

CarterFendley commented:
> Maybe worth adding just one unit test to verify if setting both env vars will generate the Workflow yaml with the new flags.

Will do :)

CarterFendley commented:
Okay I have updated the [compiler tests](https://github.com/kubeflow/pipelines/blob/714901fc4166b9bdeb629ff31d771eb49b1ef89f/backend/src/v2/compiler/argocompiler/argo_test.go#L79-L87) with logic to optional take in environment variables and set them:

```go
if tt.envVars != nil {
    for _, envVar := range tt.envVars {
	    parts := strings.Split(strings.ReplaceAll(envVar, " ", ""), "=")
	    os.Setenv(parts[0], parts[1])
    
	    // Unset after test cases has ended
	    defer os.Unsetenv(parts[0])
    }
}
```

To [test cases](https://github.com/kubeflow/pipelines/blob/714901fc4166b9bdeb629ff31d771eb49b1ef89f/backend/src/v2/compiler/argocompiler/argo_test.go#L62-L74) and [golden files](https://github.com/kubeflow/pipelines/tree/714901fc4166b9bdeb629ff31d771eb49b1ef89f/backend/src/v2/compiler/argocompiler/testdata/with_logging) have been added to test the logic included in this PR.

```go
{
	jobPath:          "../testdata/hello_world.json",
	platformSpecPath: "",
	argoYAMLPath:     "testdata/with_logging/hello_world.yaml",
	envVars:          []string{"DRIVER_LOG_LEVEL=5", "LAUNCHER_LOG_LEVEL=5"},
},
{
	jobPath:          "../testdata/importer.json",
	platformSpecPath: "",
	argoYAMLPath:     "testdata/with_logging/importer.yaml",
	envVars:          []string{"DRIVER_LOG_LEVEL=5", "LAUNCHER_LOG_LEVEL=5"},
},
```

rimolive commented:
/lgtm

CarterFendley commented on backend/src/v2/cmd/driver/main.go:
```diff
@@ -68,12 +68,17 @@ var (
 	flag.Parse()
+
+	glog.Infof("Setting log level to: '%s'", *logLevel)
+	flag.Set("v", *logLevel)
```
Good catch!

CarterFendley commented on backend/src/v2/cmd/launcher-v2/main.go:
```diff
@@ -54,6 +55,9 @@ func run() error {
 	ctx := context.Background()
 
+	glog.Infof("Setting log level to: '%s'", *logLevel)
+	flag.Set("v", *logLevel)
```
Good catch!

CarterFendley commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -303,8 +318,9 @@ func (c *workflowCompiler) addContainerExecutorTemplate(refName string) string {
 				Name:    "kfp-launcher",
-				Image:   GetLauncherImage(),
-				Command: []string{"launcher-v2", "--copy", component.KFPLauncherPath},
+				Image:   c.launcherImage,
```
No not really, just made it a bit concise to add flags and follows the pattern used in the other Container definitions for driver / launcher

hbelmiro reviewed (commented):
/lgtm

gregsheremeta commented:
I wanted to install this and run it, and also verify that it's backwards compatible. All looks good with that, but I noticed one problem. You missed one of the launcher invocations, roughly here
https://github.com/kubeflow/pipelines/blob/c2a77134265dbbe7ca8492e6e429fd3cc60b8419/backend/src/v2/driver/driver.go#L389-L405
The symptom is that even though I set LAUNCHER_LOG_LEVEL to something other than 1, the user code (impl) container logs always say `Setting log level to: '1'`

Launcher is invoked twice, in two different ways :facepalm:. 

1. First it's invoked to copy itself into the main container (you caught that invocation -- the one with `--copy` -- this runs in the `kfp-launcher` init-container), and once to actually launch the user's code (this runs in the `main` container). 
2. The second one is the one that is much more important to be able to control the logging in. That invocation is generated in the pod spec patch in driver.go. (This is all super arcane, so let me know if this isn't making sense.)

Once you add it, you should see it in the output of podSpecPatch in a container-driver log. Roughly:
```
I1018 20:34:19.156021      18 main.go:246] output podSpecPatch=
{"containers":[{"name":"main","image":"python:3.7","command":["/kfp-launcher/launch","--pipeline_name","my-pipeline","--run_id","SNIP","--execution_id","291","--executor_input","SNIP","--component_spec","SNIP","--pod_name","$(KFP_POD_NAME)","--pod_uid","$(KFP_POD_UID)","--mlmd_server_address","$(METADATA_GRPC_SERVICE_HOST)","--mlmd_server_port","$(METADATA_GRPC_SERVICE_PORT)","--", 
**(you'll see your log_level here...)**
],
```
And then the user code container log should respect the launcher log setting:
`Setting log level to: 'whatever_you_set'`

hbelmiro commented:
@CarterFendley is there anything pending in this PR?

CarterFendley commented:
@hbelmiro Yes, as noted by Greg, we want to make sure that the main container for the `-impl` pod also is setup to be configurable.

Had some discussion on slack about the possible implementations. I am OOO probably all of next week, will add that feature soon after I am back

hbelmiro commented:
@CarterFendley any updates on this?

CarterFendley commented:
@hbelmiro Thanks for the ping, I'll slack you real quick.

HumairAK commented:
@CarterFendley , just a bump, can this be rebased, and could you provide update on status?

CarterFendley commented:
Thanks for the bump, will provide update soon. Apologies for the delay.

On Fri, Jan 10, 2025 at 11:47 Humair Khan ***@***.***> wrote:

> @CarterFendley <https://github.com/CarterFendley> , just a bump, can this
> be rebased, and could you provide update on status?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/pipelines/pull/11278#issuecomment-2583235631>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEBVPQXAVIAWMFE2JG7CDT2J72TJAVCNFSM6AAAAABPS45HAOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBTGIZTKNRTGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

CarterFendley commented:
Modifications have been made to address [this issue](https://github.com/kubeflow/pipelines/pull/11278#issuecomment-2423229241) found by @gregsheremeta, thanks for pointing that out! Since the instance was one where the **driver** sets the log level of the **launcher** a design change was made to have one unified `PIPELINE_LOG_LEVEL` env var to prevent the somewhat confusing implementations of passing the `LAUNCHER_LOG_LEVEL` as a command line argument to the driver (or similar implementations).

The new usage is to update the environment variable on the `ml-pipelines` deployment to the following. 
```
    spec:
      containers:
      - env:
        - name: PIPELINE_LOG_LEVEL
          value: "3"
```
**Importantly**, as before, it is important that a numerical value such as the literal 3 not "3" here will be invalid deployment spec and validation on the spec will fail causing `kubectl edit` to reject it with the `message: error: deployments.apps "ml-pipeline"` is invalid.

After these updates, the `main` container now also runs the launcher with configured log level.
<img width="866" alt="Screenshot 2025-02-19 at 7 34 27 PM" src="https://github.com/user-attachments/assets/6e283df3-eca9-48ea-83da-ef52390e940d" />

@hbelmiro @HumairAK, or any others: Please let me know if you have any additional feedback on this PR. Apologies for the delay in the patch!

CarterFendley commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -39,6 +39,7 @@ const (
 	DefaultDriverCommand     = "driver"
 	DriverCommandEnvVar      = "V2_DRIVER_COMMAND"
 	PipelineRunAsUserEnvVar  = "PIPELINE_RUN_AS_USER"
+	PipelineLogLevelEnvVar   = "PIPELINE_LOG_LEVEL"
```
[Done](https://github.com/kubeflow/pipelines/pull/11278/commits/e6b2d654f6365bba5911c44c09cf3b5b14d1e95f)!

HumairAK commented:
/lgtm
/approve
